### DISCUSSION TITLE
How to tell `Async::IO::SSLEndpoint#connect` to not verify the certificate?

### DISCUSSION THREAD

**Discussion by postmodern** [socketry/async-io#84] [Q&A] [answered]
I'm trying to connect to an SSL port and obtain the `peer_cert`. To do this reliably, I need to ignore any certificate verification.

The documentation for [Async::IO::Endpoint.ssl](https://socketry.github.io/async-io/source/Async/IO/Endpoint/index.html) doesn't explain how to pass in additional OpenSSL options.

**Comment by postmodern** [comment]
I'm also unclear where/how `:ssl_params` is passed in to populate the `OpenSSL::SSL::SSLContext`?
https://github.com/socketry/async-io/blob/6b5dc269b4526460bbb4caa2d33cf324b686bf13/lib/async/io/ssl_endpoint.rb#L41-L46

**Reply by ioquatix** [reply]
The `async-io` gem will be archived soon, as `async` v1 is now considered EOL but still supported.

https://github.com/socketry/io-endpoint/blob/main/lib/io/endpoint/ssl_endpoint.rb#L104-L117 is the correct place to check, and I believe you could use the `ssl_params:` option. However, that option name is ugly, in fact the dependence on OpenSSL is ugly. We should probably introduce a generic "secure context" for connection and map that to `SSLContext`... there is work to be done here, especially w.r.t HTTP/3 TLS which doesn't use OpenSSL.

**Reply by postmodern** [reply]
Thanks for the heads up. Will `io-endpoint` completely replace `async-io`?

I would not rename `ssl_context` to "secure context", as that is introducing new terminology that is not canonical SSL/TLS terminology (unless I'm mistaken?). If it accepts an `OpenSSL::SSL::SSLContext` it should be called `context` or `ssl_context` or `tls_context`, etc.

**Reply by ioquatix** [reply]
We will probably accept a new "generic secure context" in the future, which can control HTTP/3 TLS as well as HTTP/2 TLS and so on. We won't remove the existing support but it might be augmented by a super set API.

**Reply by postmodern** [reply]
What would be the difference between HTTP/3 TLS and HTTP/2 TLS contexts? They both would accept a `OpenSSL::SSL::SSLContext` or a nested Hash of options that would be used to populate a `OpenSSL::SSL::SSLContext` object?

**Reply by ioquatix** [reply]
No, OpenSSL is not compatible with the HTTP/3 style encryption which works on UDP payloads rather than a TCP stream.

**Reply by postmodern** [reply]
[OpenSSL does support QUIC](https://www.openssl.org/docs/manmaster/man7/openssl-quic.html) which is what [HTTP/3 is currently building upon](https://httpwg.org/specs/rfc9114.html#rfc.section.1.2).

**Reply by ioquatix** [reply]
Nice, I didn't know they finally added support, it's literally been years.

Unfortunately, it looks like it's only client side support:

> OpenSSL 3.2 and later features support for the QUIC transport protocol. Currently, only client connectivity is supported.

**Reply by postmodern** [reply]
I'd give it time. HTTP/3 still hasn't been ratified yet, so who knows what will change in the future.

**Comment by postmodern** [comment, marked as answer]
Hmm, after passing in a custom `OpenSSL::SSL::SSLContext` with `verify_mode` set to `OpenSSL::SSL::VERIFY_NONE`, it appears that I'm actually hitting some kind of odd SSL/TLS handshake failure that also happens when using `SSLSocket` or even `curl`.

```ruby
context = OpenSSL::SSL::SSLContext.new
context.verify_mode = OpenSSL::SSL::VERIFY_NONE

host = '172.66.40.156'
port = 8443
endpoint = Async::IO::Endpoint.ssl(address,port, ssl_context: context)

endpoint.connect do |socket|
  # ...
end
```
```
OpenSSL::SSL::SSLError: SSL_connect returned=1 errno=0 peeraddr=172.66.40.156:8443 state=error: ssl/tls alert handshake failure
```
```
$ curl -k https://172.66.40.156:8443
curl: (35) OpenSSL/3.2.1: error:0A000410:SSL routines::ssl/tls alert handshake failure
```

**Reply by ioquatix** [reply]
I tried connecting with OpenSSL s_connect and couldn't get any further:

```
> openssl s_client -connect 172.66.40.156:8443 -tls1_3
Connecting to 172.66.40.156
CONNECTED(00000003)
40073167A6720000:error:0A000410:SSL routines:ssl3_read_bytes:ssl/tls alert handshake failure:ssl/record/rec_layer_s3.c:907:SSL alert number 40
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 229 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---

> openssl s_client -connect 172.66.40.156:8443 -tls1_2
Connecting to 172.66.40.156
CONNECTED(00000003)
40670791637A0000:error:0A000410:SSL routines:ssl3_read_bytes:ssl/tls alert handshake failure:ssl/record/rec_layer_s3.c:907:SSL alert number 40
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 194 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1720829726
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```
